core/src/main/scala/caliban/transformers/Transformer.scala

```diff
@@ -0,0 +1,178 @@
+package caliban.transformers
+
+import caliban.InputValue
+import caliban.introspection.adt._
+import caliban.schema.Step
+import caliban.schema.Step.{ FunctionStep, MetadataFunctionStep, ObjectStep }
+
+/**
+ * A transformer is able to modify a type, modifying its schema and the way it is resolved.
+ */
+trait Transformer[-R] { self =>
+  val typeVisitor: TypeVisitor
+
+  def transformStep[R1 <: R]: PartialFunction[Step[R1], Step[R1]]
+
+  def |+|[R0 <: R](that: Transformer[R0]): Transformer[R0] = new Transformer[R0] {
+    val typeVisitor: TypeVisitor = self.typeVisitor |+| that.typeVisitor
+
+    def transformStep[R1 <: R0]: PartialFunction[Step[R1], Step[R1]] =
+      Function.unlift { step =>
+        val modifiedStep = self.transformStep.lift(step).getOrElse(step)
+        that.transformStep.lift(modifiedStep)
+      }
```

Comment:
In case we want to keep the PartialFunction, I struggled a bit with following the composition here. Perhaps this might be a bit more readable although I'm not 100% sure whether they do the same thing:

```suggestion
    def transformStep[R1 <: R0]: PartialFunction[Step[R1], Step[R1]] = {
      def withDefault(tr: Transformer[R1]): PartialFunction[Step[R1], Step[R1]] =
        tr.transformStep[R1].orElse { case x => x }
      withDefault(self) andThen withDefault(that)
    }
```

Reply:
Yeah my code here was wrong, fixed it in my other branch, will propagate to this one.

Reply:
Fixed